### PR TITLE
Add order dataValidation

### PR DESCRIPTION
**Pull request by ahmardi**
Exception
image

### CODE REVIEW

File: Model/Data/OrderUpdate.php [hunk 1]
```diff
@@ -196,4 +196,31 @@ private function getOrderUrl(OrderInterface $order): string
     {
         return $this->storeManager->getStore()->getBaseUrl() . 'sales/order/view/order_id/' . $order->getEntityId();
     }
+
+    /**
+     * Validates the order data.
+     *
+     * @param Order $order
```
Comment (Skullsneeze) [inline Model/Data/OrderUpdate.php:203]: You're missing the param for amwal order data

File: Model/Data/OrderUpdate.php [hunk 2]
```diff
@@ -196,4 +196,31 @@ private function getOrderUrl(OrderInterface $order): string
...
+     * @return bool|string True if validation passes, otherwise returns error message.
+     */
+    private function dataValidation(Order $order, $amwalOrderData)
+    {
+        // Define an associative array mapping Amwal order fields to their corresponding getter methods.
+        // This array is used for comparing Order object values with Amwal order data.
+        $fields = ['AmwalOrderId' => 'getId', 'RefId' => 'getRefId', 'BaseGrandTotal' => 'getAmount'];
+        try {
+            foreach ($fields as $orderMethod => $amwalMethod) {
+                $orderValue = call_user_func([$order, 'get' . $orderMethod]);
```
Comment (Skullsneeze) [inline Model/Data/OrderUpdate.php:213]: I think both order and amwal order data are data objects, so we should be able to doo something like
```suggestion
                $orderValue = $order->getData($orderField);
```

Where $orderField would be the snake_case version of the fieldd

File: Model/Data/OrderUpdate.php [hunk 3]
```diff
@@ -196,4 +196,31 @@ private function getOrderUrl(OrderInterface $order): string
...
+     * Validates the order data.
+     *
+     * @param Order $order
+     * @return bool|string True if validation passes, otherwise returns error message.
+     */
+    private function dataValidation(Order $order, $amwalOrderData)
+    {
+        // Define an associative array mapping Amwal order fields to their corresponding getter methods.
+        // This array is used for comparing Order object values with Amwal order data.
+        $fields = ['AmwalOrderId' => 'getId', 'RefId' => 'getRefId', 'BaseGrandTotal' => 'getAmount'];
```
Comment (Skullsneeze) [inline Model/Data/OrderUpdate.php:210]: Might be a good idea to as this as a constant instead?

**Review by Skullsneeze — CHANGES_REQUESTED** [review]

**Comment by ahmardi** [comment]
Screenshot 2023-11-23 at 13 54 43

**Review by sameh-amwal — APPROVED** [review]

File: Model/Data/OrderUpdate.php [hunk 4]
```diff
@@ -196,4 +202,34 @@ private function getOrderUrl(OrderInterface $order): string
...
+    private function dataValidation(Order $order, DataObject $amwalOrderData)
+    {
+        // Define an associative array mapping Amwal order fields to their corresponding getter methods.
+        // This array is used for comparing Order object values with Amwal order data.
+        $fields = self::FIELD_MAPPINGS;
+        try {
+            if (floatval($order->getBaseGrandTotal()) != floatval($amwalOrderData->getAmount())) {
+                throw new \Exception(sprintf('Order (%s) %s does not match Amwal Order %s (%s != %s)', $order->getIncrementId(), 'base_grand_total', 'amount', $order->getBaseGrandTotal(), $amwalOrderData->getAmount()));
+            }
+            foreach ($fields as $orderMethod => $amwalMethod) {
```
Comment (Skullsneeze) [inline Model/Data/OrderUpdate.php:222]: You could use the constant het directly instead of singing it to its own variable first

**Review by Skullsneeze — APPROVED** [review]

**Review by Skullsneeze — APPROVED** [review]

**Pull request merged**